File: server/src/main/java/app/com/server/service/FriendshipService.java
package app.com.server.service;

import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import app.com.server.dto.CreateFriendshipDto;
import app.com.server.dto.FriendshipResponseDto;
import app.com.server.enums.FriendshipInvitationStatus;
import app.com.server.mapper.FriendshipMapper;
import app.com.server.model.Friendship;
import app.com.server.model.User;
import app.com.server.repos.FriendshipRepository;
import app.com.server.repos.UserRepository;

@Service
public class FriendshipService {
    
    @Autowired
    private FriendshipRepository friendshipRepository;
    
    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private FriendshipMapper friendshipMapper;
    
    // Send a friendship request
    public FriendshipResponseDto sendFriendshipRequest(UUID senderId, CreateFriendshipDto createFriendshipDto) {
        // Validate input
        if (createFriendshipDto.getReceiverEmail() == null || createFriendshipDto.getReceiverEmail().trim().isEmpty()) {
            throw new IllegalArgumentException("Receiver email is required");
        }
        
        // Find sender
        Optional<User> senderOpt = userRepository.findById(senderId);
        if (senderOpt.isEmpty()) {
            throw new IllegalArgumentException("Sender not found");
        }
        User sender = senderOpt.get();
        
        // Find receiver by email
        Optional<User> receiverOpt = userRepository.findByEmail(createFriendshipDto.getReceiverEmail());
        if (receiverOpt.isEmpty()) {
            throw new IllegalArgumentException("Receiver not found with email: " + createFriendshipDto.getReceiverEmail());
        }
        User receiver = receiverOpt.get();
        
        // Check if sender and receiver are the same
        if (sender.getId().equals(receiver.getId())) {
            throw new IllegalArgumentException("Cannot send friendship request to yourself");
        }
        
        // Check if friendship already exists
        if (friendshipRepository.existsFriendshipBetweenUsers(senderId, receiver.getId())) {
            throw new IllegalArgumentException("Friendship already exists between these users");
        }
        
        // Create friendship request
        Friendship friendship = friendshipMapper.toEntity(createFriendshipDto, sender, receiver);
        Friendship savedFriendship = friendshipRepository.save(friendship);
        
        return friendshipMapper.toResponseDto(savedFriendship);
    }
    
    // Accept a friendship request
    public FriendshipResponseDto acceptFriendshipRequest(UUID friendshipId, UUID userId) {
        Optional<Friendship> friendshipOpt = friendshipRepository.findById(friendshipId);
        if (friendshipOpt.isEmpty()) {
            throw new IllegalArgumentException("Friendship request not found");
        }
        
        Friendship friendship = friendshipOpt.get();
        
        // Check if user is the receiver
        if (!friendship.getReceiver().getId().equals(userId)) {
            throw new IllegalArgumentException("You can only accept friendship requests sent to you");
        }
        
        // Check if status is pending
        if (friendship.getStatus() != FriendshipInvitationStatus.PENDING) {
            throw new IllegalArgumentException("This friendship request is not pending");
        }
        
        // Update status to accepted
        friendship.setStatus(FriendshipInvitationStatus.ACCEPTED);
        friendship.setUpdatedAt(Instant.now());
        
        Friendship savedFriendship = friendshipRepository.save(friendship);
        return friendshipMapper.toResponseDto(savedFriendship);
    }
    
    // Reject a friendship request
    public void rejectFriendshipRequest(UUID friendshipId, UUID userId) {
        Optional<Friendship> friendshipOpt = friendshipRepository.findById(friendshipId);
        if (friendshipOpt.isEmpty()) {
            throw new IllegalArgumentException("Friendship request not found");
        }
        
        Friendship friendship = friendshipOpt.get();
        
        // Check if user is the receiver
        if (!friendship.getReceiver().getId().equals(userId)) {
            throw new IllegalArgumentException("You can only reject friendship requests sent to you");
        }
        
        // Delete the friendship request
        friendshipRepository.delete(friendship);
    }
    
    // Cancel a friendship request (for sender)
    public void cancelFriendshipRequest(UUID friendshipId, UUID userId) {
        Optional<Friendship> friendshipOpt = friendshipRepository.findById(friendshipId);
        if (friendshipOpt.isEmpty()) {
            throw new IllegalArgumentException("Friendship request not found");
        }
        
        Friendship friendship = friendshipOpt.get();
        
        // Check if user is the sender
        if (!friendship.getSender().getId().equals(userId)) {
            throw new IllegalArgumentException("You can only cancel friendship requests sent by you");
        }
        
        // Check if status is pending
        if (friendship.getStatus() != FriendshipInvitationStatus.PENDING) {
            throw new IllegalArgumentException("You can only cancel pending friendship requests");
        }
        
        // Delete the friendship request
        friendshipRepository.delete(friendship);
    }
    
    // Block a user
    public FriendshipResponseDto blockUser(UUID userId, UUID userToBlockId) {
        // Check if users exist
        Optional<User> userOpt = userRepository.findById(userId);
        Optional<User> userToBlockOpt = userRepository.findById(userToBlockId);
        
        if (userOpt.isEmpty() || userToBlockOpt.isEmpty()) {
            throw new IllegalArgumentException("User not found");
        }
        
        User user = userOpt.get();
        User userToBlock = userToBlockOpt.get();
        
        // Check if trying to block yourself
        if (userId.equals(userToBlockId)) {
            throw new IllegalArgumentException("Cannot block yourself");
        }
        
        // Check if friendship already exists
        Optional<Friendship> existingFriendship = friendshipRepository.findBySenderIdAndReceiverId(userId, userToBlockId);
        if (existingFriendship.isEmpty()) {
            existingFriendship = friendshipRepository.findBySenderIdAndReceiverId(userToBlockId, userId);
        }
        
        Friendship friendship;
        if (existingFriendship.isPresent()) {
            friendship = existingFriendship.get();
            friendship.setStatus(FriendshipInvitationStatus.BLOCKED);
            friendship.setUpdatedAt(Instant.now());
        } else {
            friendship = new Friendship();
            friendship.setSender(user);
            friendship.setReceiver(userToBlock);
            friendship.setStatus(FriendshipInvitationStatus.BLOCKED);
            friendship.setCreatedAt(Instant.now());
            friendship.setUpdatedAt(Instant.now());
        }
        
        Friendship savedFriendship = friendshipRepository.save(friendship);
        return friendshipMapper.toResponseDto(savedFriendship);
    }
    
    // Unblock a user
    public void unblockUser(UUID userId, UUID userToUnblockId) {
        Optional<Friendship> friendshipOpt = friendshipRepository.findBySenderIdAndReceiverId(userId, userToUnblockId);
        if (friendshipOpt.isEmpty()) {
            friendshipOpt = friendshipRepository.findBySenderIdAndReceiverId(userToUnblockId, userId);
        }
        
        if (friendshipOpt.isEmpty()) {
            throw new IllegalArgumentException("No friendship found between these users");
        }
        
        Friendship friendship = friendshipOpt.get();
        if (friendship.getStatus() != FriendshipInvitationStatus.BLOCKED) {
            throw new IllegalArgumentException("User is not blocked");
        }
        
        // Delete the friendship (unblock)
        friendshipRepository.delete(friendship);
    }
    
    // Remove a friend
    public void removeFriend(UUID userId, UUID friendId) {
        Optional<Friendship> friendshipOpt = friendshipRepository.findBySenderIdAndReceiverId(userId, friendId);
        if (friendshipOpt.isEmpty()) {
            friendshipOpt = friendshipRepository.findBySenderIdAndReceiverId(friendId, userId);
        }
        
        if (friendshipOpt.isEmpty()) {
            throw new IllegalArgumentException("No friendship found between these users");
        }
        
        Friendship friendship = friendshipOpt.get();
        if (friendship.getStatus() != FriendshipInvitationStatus.ACCEPTED) {
            throw new IllegalArgumentException("Users are not friends");
        }
        
        // Delete the friendship
        friendshipRepository.delete(friendship);
    }
    
    // Get all friends of a user
    public List<FriendshipResponseDto> getFriends(UUID userId) {
        List<Friendship> friendships = friendshipRepository.findAcceptedFriendshipsByUserId(userId);
        return friendships.stream()
                .map(friendshipMapper::toResponseDto)
                .collect(Collectors.toList());
    }
    
    // Get pending friendship requests received by user
    public List<FriendshipResponseDto> getPendingRequests(UUID userId) {
        List<Friendship> friendships = friendshipRepository.findPendingFriendshipRequests(userId);
        return friendships.stream()
                .map(friendshipMapper::toResponseDto)
                .collect(Collectors.toList());
    }
    
    // Get sent friendship requests by user
    public List<FriendshipResponseDto> getSentRequests(UUID userId) {
        List<Friendship> friendships = friendshipRepository.findSentFriendshipRequests(userId);
        return friendships.stream()
                .map(friendshipMapper::toResponseDto)
                .collect(Collectors.toList());
    }
    
    // Get all friendships of a user (friends, pending, blocked)
    public List<FriendshipResponseDto> getAllFriendships(UUID userId) {
        List<Friendship> friendships = friendshipRepository.findByUserId(userId);
        return friendships.stream()
                .map(friendshipMapper::toResponseDto)
                .collect(Collectors.toList());
    }
    
    // Get friendship status between two users
    public FriendshipResponseDto getFriendshipStatus(UUID userId1, UUID userId2) {
        Optional<Friendship> friendshipOpt = friendshipRepository.findBySenderIdAndReceiverId(userId1, userId2);
        if (friendshipOpt.isEmpty()) {
            friendshipOpt = friendshipRepository.findBySenderIdAndReceiverId(userId2, userId1);
        }
        
        if (friendshipOpt.isEmpty()) {
            return null; // No friendship exists
        }
        
        return friendshipMapper.toResponseDto(friendshipOpt.get());
    }
    
    // Check if two users are friends
    public boolean areFriends(UUID userId1, UUID userId2) {
        return friendshipRepository.findAcceptedFriendshipsByUserId(userId1).stream()
                .anyMatch(f -> f.getSender().getId().equals(userId2) || f.getReceiver().getId().equals(userId2));
    }
    
    // Get friendship by ID
    public FriendshipResponseDto getFriendshipById(UUID friendshipId) {
        Optional<Friendship> friendshipOpt = friendshipRepository.findById(friendshipId);
        if (friendshipOpt.isEmpty()) {
            throw new IllegalArgumentException("Friendship not found");
        }
        
        return friendshipMapper.toResponseDto(friendshipOpt.get());
    }
}

File: front/src/app/core/services/friendship.service.ts
import { inject, Injectable } from '@angular/core';
import {
  Friendship,
  CreateFriendshipRequest,
  FriendshipStatus,
  FriendshipState,
} from '../models/friendship.model';
import { HttpClient, HttpParams } from '@angular/common/http';
import {
  BehaviorSubject,
  catchError,
  finalize,
  map,
  Observable,
  of,
  tap,
  throwError,
} from 'rxjs';
import { AppConfig } from '../config/app.config';
import { ChatService } from './chat.service';

@Injectable({
  providedIn: 'root',
})
export class FriendshipService {
  private http = inject(HttpClient);
  private chatService = inject(ChatService);
  private state = new BehaviorSubject<FriendshipState>({
    friendships: [],
    friends: [],
    pendingRequests: [],
    sentRequests: [],
    loading: false,
    error: null,
  });

  state$ = this.state.asObservable();
  friendships$ = this.state$.pipe(
    map((state: FriendshipState) => state.friendships)
  );
  friends$ = this.state$.pipe(map((state: FriendshipState) => state.friends));
  pendingRequests$ = this.state$.pipe(
    map((state: FriendshipState) => state.pendingRequests)
  );
  sentRequests$ = this.state$.pipe(
    map((state: FriendshipState) => state.sentRequests)
  );
  loading$ = this.state$.pipe(map((state: FriendshipState) => state.loading));
  error$ = this.state$.pipe(map((state: FriendshipState) => state.error));

  constructor() {}

  // Send friendship request
  sendFriendshipRequest(
    senderId: string,
    request: CreateFriendshipRequest
  ): Observable<Friendship> {
    this.setLoading(true);
    const params = new HttpParams().set('senderId', senderId);

    return this.http
      .post<Friendship>(AppConfig.FRIENDSHIP_ENDPOINTS.SEND_REQUEST, request, {
        params,
      })
      .pipe(
        tap((friendship) => {
          this.updateState({
            sentRequests: [...this.state.value.sentRequests, friendship],
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to send friendship request');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Accept friendship request
  acceptFriendshipRequest(
    friendshipId: string,
    userId: string
  ): Observable<Friendship> {
    this.setLoading(true);
    const params = new HttpParams().set('userId', userId);

    return this.http
      .put<Friendship>(
        `${AppConfig.FRIENDSHIP_ENDPOINTS.ACCEPT_REQUEST}/${friendshipId}/accept`,
        {},
        { params }
      )
      .pipe(
        tap((friendship) => {
          const currentState = this.state.value;
          const updatedPending = currentState.pendingRequests.filter(
            (f) => f.id !== friendshipId
          );
          const updatedFriends = [...currentState.friends, friendship];

          this.updateState({
            pendingRequests: updatedPending,
            friends: updatedFriends,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to accept friendship request');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Reject friendship request
  rejectFriendshipRequest(
    friendshipId: string,
    userId: string
  ): Observable<void> {
    this.setLoading(true);
    const params = new HttpParams().set('userId', userId);

    return this.http
      .delete<void>(
        `${AppConfig.FRIENDSHIP_ENDPOINTS.REJECT_REQUEST}/${friendshipId}/reject`,
        { params }
      )
      .pipe(
        tap(() => {
          const currentState = this.state.value;
          const updatedPending = currentState.pendingRequests.filter(
            (f) => f.id !== friendshipId
          );

          this.updateState({
            pendingRequests: updatedPending,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to reject friendship request');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Cancel friendship request
  cancelFriendshipRequest(
    friendshipId: string,
    userId: string
  ): Observable<void> {
    this.setLoading(true);
    const params = new HttpParams().set('userId', userId);
    return this.http
      .delete<void>(
        `${AppConfig.FRIENDSHIP_ENDPOINTS.CANCEL_REQUEST}/${friendshipId}/cancel`,
        { params }
      )
      .pipe(
        tap(() => {
          const currentState = this.state.value;
          const updatedPending = currentState.pendingRequests.filter(
            (f) => f.id !== friendshipId
          );
          this.updateState({
            pendingRequests: updatedPending,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to cancel friendship request');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Remove friend
  removeFriend(userId: string, friendId: string): Observable<void> {
    this.setLoading(true);
    const params = new HttpParams()
      .set('userId', userId)
      .set('friendId', friendId);

    return this.http
      .delete<void>(AppConfig.FRIENDSHIP_ENDPOINTS.REMOVE_FRIEND, { params })
      .pipe(
        tap(() => {
          const currentState = this.state.value;
          const updatedFriends = currentState.friends.filter(
            (f) => f.sender.id !== friendId && f.receiver.id !== friendId
          );

          this.updateState({
            friends: updatedFriends,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to remove friend');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Block user
  blockUser(userId: string, userToBlockId: string): Observable<Friendship> {
    this.setLoading(true);
    const params = new HttpParams()
      .set('userId', userId)
      .set('userToBlockId', userToBlockId);

    return this.http
      .post<Friendship>(
        AppConfig.FRIENDSHIP_ENDPOINTS.BLOCK_USER,
        {},
        { params }
      )
      .pipe(
        tap((friendship) => {
          const currentState = this.state.value;
          const updatedFriends = currentState.friends.filter(
            (f) =>
              f.sender.id !== userToBlockId && f.receiver.id !== userToBlockId
          );

          this.updateState({
            friends: updatedFriends,
            friendships: [
              ...currentState.friendships.filter((f) => f.id !== friendship.id),
              friendship,
            ],
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to block user');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Unblock user
  unblockUser(userId: string, userToUnblockId: string): Observable<void> {
    this.setLoading(true);
    const params = new HttpParams()
      .set('userId', userId)
      .set('userToUnblockId', userToUnblockId);

    return this.http
      .delete<void>(AppConfig.FRIENDSHIP_ENDPOINTS.UNBLOCK_USER, { params })
      .pipe(
        tap(() => {
          const currentState = this.state.value;
          const updatedFriendships = currentState.friendships.filter(
            (f) =>
              !(
                f.sender.id === userToUnblockId ||
                f.receiver.id === userToUnblockId
              ) || f.status !== FriendshipStatus.BLOCKED
          );

          this.updateState({
            friendships: updatedFriendships,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to unblock user');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Get all friends
  getFriends(userId: string): Observable<Friendship[]> {
    this.setLoading(true);
    const params = new HttpParams().set('userId', userId);

    return this.http
      .get<Friendship[]>(AppConfig.FRIENDSHIP_ENDPOINTS.GET_FRIENDS, { params })
      .pipe(
        tap((friends) => {
          this.updateState({
            friends,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to load friends');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Get pending friendship requests
  getPendingRequests(userId: string): Observable<Friendship[]> {
    this.setLoading(true);
    const params = new HttpParams().set('userId', userId);

    return this.http
      .get<Friendship[]>(AppConfig.FRIENDSHIP_ENDPOINTS.GET_PENDING, { params })
      .pipe(
        tap((pendingRequests) => {
          this.updateState({
            pendingRequests,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to load pending requests');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Get sent friendship requests
  getSentRequests(userId: string): Observable<Friendship[]> {
    this.setLoading(true);
    const params = new HttpParams().set('userId', userId);

    return this.http
      .get<Friendship[]>(AppConfig.FRIENDSHIP_ENDPOINTS.GET_SENT, { params })
      .pipe(
        tap((sentRequests) => {
          this.updateState({
            sentRequests,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to load sent requests');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Get all friendships
  getAllFriendships(userId: string): Observable<Friendship[]> {
    this.setLoading(true);
    const params = new HttpParams().set('userId', userId);

    return this.http
      .get<Friendship[]>(AppConfig.FRIENDSHIP_ENDPOINTS.GET_ALL, { params })
      .pipe(
        tap((friendships) => {
          this.updateState({
            friendships,
            loading: false,
            error: null,
          });
        }),
        catchError((error) => {
          this.setError('Failed to load friendships');
          return throwError(() => error);
        }),
        finalize(() => this.setLoading(false))
      );
  }

  // Get friendship status between two users
  getFriendshipStatus(
    userId1: string,
    userId2: string
  ): Observable<Friendship | null> {
    const params = new HttpParams()
      .set('userId1', userId1)
      .set('userId2', userId2);

    return this.http
      .get<Friendship>(AppConfig.FRIENDSHIP_ENDPOINTS.GET_STATUS, { params })
      .pipe(
        catchError((error) => {
          if (error.status === 404) {
            return of(null);
          }
          this.setError('Failed to get friendship status');
          return throwError(() => error);
        })
      );
  }

  // Check if two users are friends
  areFriends(userId1: string, userId2: string): Observable<boolean> {
    const params = new HttpParams()
      .set('userId1', userId1)
      .set('userId2', userId2);

    return this.http
      .get<boolean>(AppConfig.FRIENDSHIP_ENDPOINTS.CHECK_FRIENDS, { params })
      .pipe(
        catchError((error) => {
          this.setError('Failed to check friendship status');
          return throwError(() => error);
        })
      );
  }

  // Get friendship by ID
  getFriendshipById(friendshipId: string): Observable<Friendship> {
    return this.http
      .get<Friendship>(
        `${AppConfig.FRIENDSHIP_ENDPOINTS.GET_BY_ID}/${friendshipId}`
      )
      .pipe(
        catchError((error) => {
          this.setError('Failed to get friendship');
          return throwError(() => error);
        })
      );
  }

  // Refresh all friendship data for a user
  refreshFriendshipData(userId: string): Observable<void> {
    this.setLoading(true);

    return this.getAllFriendships(userId).pipe(
      tap(() => {
        // Load friends, pending, and sent requests in parallel
        this.getFriends(userId).subscribe();
        this.getPendingRequests(userId).subscribe();
        this.getSentRequests(userId).subscribe();
      }),
      map(() => void 0),
      catchError((error) => {
        this.setError('Failed to refresh friendship data');
        return throwError(() => error);
      }),
      finalize(() => this.setLoading(false))
    );
  }

  // Helper methods
  private setLoading(loading: boolean): void {
    this.updateState({ loading });
  }

  private setError(error: string): void {
    this.updateState({ error, loading: false });
  }

  private updateState(partialState: Partial<FriendshipState>): void {
    this.state.next({ ...this.state.value, ...partialState });
  }

  // Clear error
  clearError(): void {
    this.updateState({ error: null });
  }

  // Reset state
  resetState(): void {
    this.state.next({
      friendships: [],
      friends: [],
      pendingRequests: [],
      sentRequests: [],
      loading: false,
      error: null,
    });
  }

  /**
   * Start a chat session with a friend
   */
  startChatWithFriend(
    friendId: string,
    currentUserId: string
  ): Observable<any> {
    return this.chatService.createChatSession(currentUserId, friendId);
  }
}
